numpy/lib/function_base.py

```diff
@@ -1242,11 +1242,11 @@ def gradient(f, *varargs, **kwargs):
         # Needs to keep the specific units, can't be a general unit
         otype = f.dtype
 
-    # Convert datetime64 data into ints. Make dummy variable `y`
-    # that is a view of ints if the data is datetime64, otherwise
+    # Convert datetime64 data into timedelta. Make dummy variable `y`
+    # that is a view of timedelta if the data is datetime64, otherwise
     # just set y equal to the the array `f`.
     if f.dtype.char in ["M", "m"]:
-        y = f.view('int64')
+        y = f.view(otype)
```

Comment:
there is some weird problems in gradient that made it work only if it did the wrong, fancy, indexing. I guess this maybe somehow fixed/worked around it, it is possible that the commit message explains more, but likely not.